tests/phpunit/api/v3/OrderTest.php

```diff
@@ -488,6 +488,24 @@ public function testDeleteOrder(): void {
     }
   }
 
+  /**
+   * Test order api treats amount as inclusive when line items not set.
+   *
+   * @throws \CRM_Core_Exception
+   */
+  public function testAPIOrderTax(): void {
+    $this->enableTaxAndInvoicing();
+    $this->createFinancialTypeWithSalesTax();
+    $order = $this->callAPISuccess('Order', 'create', [
+      'total_amount' => 10,
+      'financial_type_id' => 'Test taxable financial Type',
+      'contact_id' => $this->individualCreate(),
+      'sequential' => 1,
+    ])['values'][0];
+    $this->assertEquals(10, $order['total_amount']);
+    $this->assertEquals(0.47619047619048, $order['tax_amount']);
+  }
+
```

Comment:
Ah I see -> so if we're passing in $10 and $0.50 e.g. The $10 and $0.50

But if the financial type is of a type that has Sales Tax associated with it then we need to API to calculate sales tax for us. And that then becomes 0.47619047619048 ✅

Reply:
@KarinG right - the assumption is that the tax amount can always be calculated from the financial type & the function calling order api just passes the total amount (inclusive!) - so tax_amount is actually a deprecated parameter for the order api

Reply:
That would be a change in what the API has been doing through 5.40 though. We would have to (and could) rewrite the way we have been using Order API in the workflow. But it would be a change - (that others may hit as well)

End result for our (complicated scenario - mix and match) test is actually adding line_items ad-hoc:

Your edit to this single line -> 
'line_total' => (float) $params['total_amount'],

Makes the API work as it did before ✅

<img width="1138" alt="image" src="https://user-images.githubusercontent.com/5340555/130169466-9005fc4d-2b1b-4ed5-884d-c7569282675b.png">

Why change more things? I'm a bit worried about calculating things that don't need to be calculated - if I'm passing sales tax into the API I would expect it to return as is. 

I've always looked at Order API as creating the Order (Contribution ID) so that we can then attempt the Payment and then attach the Payment (processor just wants total_amount). Then we figure out the line items/details and write them to the record and decide who needs registrations and/or membership edits. Now some of that logic appears to be shifting to the Order API. 

E.g. note that in the example screenshot -> Main Financial Type (at the Contribution level) is Donation - so 0% Sales Tax. Yet it contains a line_item that are subject to Sales Tax. That's a real scenario -> making a Donation so that Contribution at the highest level is deductible for tax purposes - but adding a purchase to transaction.

Reply:
I'll be checking in with J tomorrow. I'll ask him to look at our logic/use of Order API as well -> this is what Matt and I ended up with as a straight up replacement for Transact API -> 

```
  function wf_civicrm_api3_contribution_transact($params) {
    // Start with the same parameters as Contribution.transact.
    $params['contribution_status_id'] = 'Pending';
    if (!isset($params['invoice_id']) && !isset($params['invoiceID'])) {
      // Set an invoice_id here if you have not already done so.
      // Potentially Order api should do this https://lab.civicrm.org/dev/financial/issues/78
      $params['invoice_id'] = md5(uniqid(rand(), TRUE));
    }
    if (isset($params['invoice_id']) && !isset($params['invoiceID'])) {
      // This would be required prior to https://lab.civicrm.org/dev/financial/issues/77
      $params['invoiceID'] = $params['invoice_id'];
    }
    elseif (!isset($params['invoice_id']) && isset($params['invoiceID'])) {
      $params['invoice_id'] = $params['invoiceID'];
    }

    $order = civicrm_api3('Order', 'create', $params);
    try {
      // Use the Payment Processor to attempt to take the actual payment. You may
      // pass in other params here, too.
      $propertyBag = new \Civi\Payment\PropertyBag();
      $propertyBag->mergeLegacyInputParams($params);
      $propertyBag->setContributionID($order['id']);
      // @fixme: Class \Civi\Payment\CRM_Utils_Money not found (fixed in 5.28 via https://github.com/civicrm/civicrm-core/pull/17505
      //   But note we still return a localized format
      // $propertyBag->setAmount($params['total_amount']);

      // We need the payment params as an array to pass to PaymentProcessor.Pay API3.
      // If https://github.com/civicrm/civicrm-core/pull/17507 was merged we wouldn't need this hack
      //   using ReflectionClass to access the array.
      $reflectionClass = new \ReflectionClass($propertyBag);
      $reflectionProperty = $reflectionClass->getProperty('props');
      $reflectionProperty->setAccessible(TRUE);
      $payParams = $reflectionProperty->getValue($propertyBag)['default'];

      // propertyBag has 'contributionID', we need 'contribution_id'.
      $payParams['contribution_id'] = $propertyBag->getContributionID();
      // propertyBag has 'currency', iATS uses `currencyID` until https://github.com/iATSPayments/com.iatspayments.civicrm/pull/350 is merged
      $payParams['currencyID'] = $propertyBag->getCurrency();
      $payParams['amount'] = $params['total_amount'];
      $payResult = reset(civicrm_api3('PaymentProcessor', 'pay', $payParams)['values']);

      // webform_civicrm sends out receipts using Contribution.send_confirmation API if the contribution page is has is_email_receipt = TRUE.
      // We allow this to be overridden here but default to FALSE.
      $params['is_email_receipt'] = $params['is_email_receipt'] ?? FALSE;

      // Assuming the payment was taken, record it which will mark the Contribution
      // as Completed and update related entities.
      civicrm_api3('Payment', 'create', [
        'contribution_id' => $order['id'],
        'total_amount' => $payParams['amount'],
        'payment_instrument_id' => $order['values'][$order['id']]['payment_instrument_id'],
        // If there is a processor, provide it:
        'payment_processor_id' => $payParams['payment_processor_id'],
        'is_send_contribution_notification' => $params['is_email_receipt'],
        'trxn_id' => $payResult['trxn_id'] ?? NULL,
      ]);
    }
    catch (\Exception $e) {
      return ['error_message' => $e->getMessage()];
    }

    $contribution = civicrm_api3('Contribution', 'getsingle', ['id' => $order['id']]);
    // Transact API works a bit differently because we are actually adding the payment and updating the contribution status.
    // A "normal" transaction (eg. using doPayment() or just PaymentProcessor.pay) would not update the contribution and would
    // rely on the caller to do so. The caller relies on the `payment_status_id` (Pending or Completed) to know whether the payment
    // was successful or not.
    $contribution['payment_status_id'] = $contribution['contribution_status_id'];
    return $contribution;
  }
```

Reply:
@KarinG the main reason for changing things was that there were various places where financials were winding up incorrect & the schema is supposed to be that the financial type id dictates the tax rate (& I don't think it's reliable without). Note this PR does not make that change

Regarding your scenario - aren't you talking about a case where you are passing in more than one line items & wouldn't the tax-attractiving line item have a tax-attracting financial type? It is per line item

Reply:
In the example test/scenario -> the Order is created with $10 and $0.25 tax (on the Contribution level) and a non-tax-attracting financial type at the highest level - we've never passed in any line_items at the stage of either Transact API or now Order API. We could start doing that -> we are doing our own tally of line items to get to the correct amounts in the first place $10 and $0.25 (obviously).

Reply:
@KarinG yeah - if you want order api to work with multiple line items the expectation is that you pass in the line items - passing in total_amount is really only for single line items

Reply:
Note I'm not sure which version but I have made it smarter about requiring less keys - this is from 5.41 tests but I think it figures out some of those values if not passed 

```
 $params = [
      'contact_id' => $this->_individualId,
      'receive_date' => '2018-01-01',
      'total_amount' => 50,
      'financial_type_id' => $this->_financialTypeId,
      'contribution_status_id' => 'Pending',
      'line_items' => [
        0 => [
          'line_item' => [
            '0' => [
              'price_field_id' => 1,
              'price_field_value_id' => 1,
              'label' => 'Test 1',
              'field_title' => 'Test 1',
              'qty' => 1,
              'unit_price' => 40,
              'line_total' => 40,
              'financial_type_id' => 1,
              'entity_table' => 'civicrm_contribution',
            ],
          ],
        ],
      ],
    ];
```